src/hazelcore/dynamics/Evaluator.re

```diff
@@ -69,258 +69,341 @@ let eval_bin_float_op =
   };
 };
 
-let rec evaluate = (d: DHExp.t): result =>
+let rec evaluate =
+        (d: DHExp.t, assert_map: AssertMap.t): (result, AssertMap.t) =>
```

Comment:
It seems `assert_map` isn't being read anywhere here. Should that be the case? If so, I'd re-write this so that `evaluate` doesn't take `assert_map` as an argument and just aggregates the submaps of children. This makes it more amenable to memoization in the future by preserving independence of evaluation of children.

Reply:
I will talk to Cyrus about this and try to find other way to get the assert map

Reply:
Not sure what you mean by that. My suggestion is for `evaluate` to have the type signature `DHExp.t => (result, AssertMap.t)`.

Reply:
Oh I see what you mean, I initially thought this caused the problem when evaluate was recursively called, but now that I think about it I think we can get rid of it.